### PR TITLE
:memo: Parallelize the Config Policy controller E2E tests

### PR DESCRIPTION
**Pull request by yiraeChristineKim**
The Config Policy controller's E2E tests in its repo takes roughly 45 minutes to run. This is due to all the test running serially. These could be optimized to run in parallel for faster execution.

Acceptance Criteria:

The config-policy-controller E2E tests run in under 20 minutes or Christine declares it impossible.

### CODE REVIEW

**Review by JustinKuli — COMMENTED** [review]
Looks good overall! Just a few small questions. I'd also prefer to see a more descriptive commit message.

File: test/resources/case28_evaluation_metric/case28_policy.yaml [hunk 1]
```diff
@@ -11,7 +11,7 @@ spec:
     - complianceType: musthave
       objectDefinition:
         apiVersion: v1
-        kind: ConfigMap
+        kind: case28-ConfigMap
```
[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply test/resources/case28_evaluation_metric/case28_policy.yaml:14]: good catch!!

File: Makefile [hunk 2]
```diff
@@ -205,8 +205,8 @@ install-resources:
 
 .PHONY: e2e-test
 e2e-test: e2e-dependencies
-	$(GINKGO) -v --fail-fast $(E2E_TEST_ARGS) test/e2e
-
+	$(GINKGO) -v -p --fail-fast -procs=7 $(E2E_TEST_ARGS) test/e2e
+#
```
[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply Makefile]: cuz our CI default is 2

File: test/e2e/case11_apiserver_config_test.go [hunk 3]
```diff
@@ -24,8 +24,8 @@ const (
 	tlsProfileInformYaml             string = "../resources/case11_apiserver_config/tls_profile_inform.yaml"
 )
 
-var _ = Describe("Test APIServer Config policy", func() {
-	Describe("Test etcd encryption and tls profile", func() {
+var _ = Describe("Test APIServer Config policy", Serial, func() {
```
Comment (yiraeChristineKim) [inline test/e2e/case11_apiserver_config_test.go:27]: This test is removing API so affect others

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply Makefile]: i wonder limitation looks more procs faster also I didn't request a review because I am not finished yet so weird.. I cannot even remove you from the reviewer list. you are the default reviewer now

File: Makefile [hunk 4]
```diff
@@ -205,7 +205,7 @@ install-resources:
 
 .PHONY: e2e-test
 e2e-test: e2e-dependencies
-	$(GINKGO) -v --fail-fast $(E2E_TEST_ARGS) test/e2e
+	$(GINKGO) -v -p -procs=7 --fail-fast $(E2E_TEST_ARGS) test/e2e
```
Comment (yiraeChristineKim) [inline Makefile:208]: CI default is 2 so I had to do this

File: test/e2e/case13_templatization_test.go [hunk 5]
```diff
@@ -256,123 +264,124 @@ var _ = Describe("Test templatization", func() {
...
 		})
-		It("Cleans up", func() {
+		AfterAll(func() {
 			deleteConfigPolicies([]string{case13Unterminated, case13WrongArgs})
 		})
 	})
 	// Though the Bugzilla bug #2007575 references a different incorrect behavior, it's the same
 	// underlying bug and this behavior is easier to test.
-	Describe("RHBZ#2007575: Test that the template updates when a referenced resource object is updated", func() {
-		const configMapName = "configmap-update-referenced-object"
```
Comment (yiraeChristineKim) [inline test/e2e/case13_templatization_test.go:266]: I just added Describe around describe

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (dhaiducek) [reply Makefile]: > I wonder limitation looks more procs faster also I didn't request a review because I am not finished yet so weird.. I cannot even remove you from the reviewer list. you are the default reviewer now

FYI for next time, if you open a Draft PR, it doesn't get assigned reviewers but I think workflows still run. 🙂

[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply Makefile]: oh!! Draft PR I will ask you next time Thanks for letting know good info!!! @dhaiducek

File: .github/workflows/kind.yml
Comment (dhaiducek) [inline .github/workflows/kind.yml:1]: Curiosity question: Are these just IDE formatting updates?

File: test/e2e/case13_templatization_test.go [hunk 6]
```diff
@@ -105,6 +105,8 @@ var _ = Describe("Test templatization", func() {
 
 				return utils.GetComplianceState(managedPlc)
 			}, defaultTimeoutSeconds, 1).Should(Equal("Compliant"))
+
+			By("Clean up")
```
Comment (dhaiducek) [inline test/e2e/case13_templatization_test.go:109]: Should this be in an `AfterAll()` function?

File: test/e2e/case13_templatization_test.go [hunk 7]
```diff
@@ -145,6 +147,8 @@ var _ = Describe("Test templatization", func() {
...
```
Comment (dhaiducek) [inline test/e2e/case13_templatization_test.go:151]: Should this be in an `AfterAll()` function?

File: test/e2e/case13_templatization_test.go [hunk 8]
```diff
@@ -194,6 +198,8 @@ var _ = Describe("Test templatization", func() {
...
```
Comment (dhaiducek) [inline test/e2e/case13_templatization_test.go:202]: Should this be in an `AfterAll()` function?

File: test/e2e/case6_no_ns_test.go [hunk 9]
```diff
@@ -35,6 +36,13 @@ var _ = Describe("Test multiple obj template handling", func() {
 
 				return utils.GetComplianceState(managedPlc)
 			}, defaultTimeoutSeconds, 1).Should(Equal("NonCompliant"))
+
+			By("Clean up")
```
Comment (dhaiducek) [inline test/e2e/case6_no_ns_test.go:40]: Should this be in an `AfterAll()` function?

File: test/e2e/case6_no_ns_test.go [hunk 10]
```diff
@@ -64,15 +72,18 @@ var _ = Describe("Test multiple obj template handling", func() {
 			ns2 := utils.GetClusterLevelWithTimeout(clientManagedDynamic, gvrNS,
 				case6NSName2, true, defaultTimeoutSeconds)
 			Expect(ns2).NotTo(BeNil())
-		})
-		It("Cleans up", func() {
+
+			By("Clean up")
```
Comment (dhaiducek) [inline test/e2e/case6_no_ns_test.go:76]: Should this be in an `AfterAll()` function?

File: test/e2e/case25_related_object_metric_test.go [hunk 11]
```diff
@@ -58,18 +57,26 @@ var _ = Describe("Test related object metrics", Ordered, func() {
...
+		)
+		utils.GetWithTimeout(
+			clientManagedDynamic, gvrConfigPolicy, policy2Name, testNamespace, false, defaultTimeoutSeconds,
+		)
 		utils.GetWithTimeout(
 			clientManagedDynamic, gvrConfigMap, relatedObject, "default", false, defaultTimeoutSeconds)
 	}
 
-	It("should clean up", cleanup)
-	It("should have no common related object metrics after clean up", func() {
```
Comment (dhaiducek) [inline test/e2e/case25_related_object_metric_test.go:69]: I don't think these lines should be removed. Adding an `AfterAll()` makes sense so that the cleanup always happens, but the cleanup before checking for metrics was intentional and part of the test.

File: test/utils/utils.go [hunk 12]
```diff
@@ -112,7 +112,7 @@ func GetWithTimeout(
 		}
 
 		return nil
-	}, timeout, 1).ShouldNot(HaveOccurred())
+	}, timeout, 1).Should(BeNil())
```
Comment (dhaiducek) [inline test/utils/utils.go:115]: Why was this changed? I thinking using `ShouldNot(HaveOccurred())` with errors might be a best practice.

**Review by dhaiducek — COMMENTED** [review]
Thanks! This is going to be so great to not have to wait an hour for the E2E tests! I have some comments/questions.

File: .github/workflows/kind.yml
Comment (yiraeChristineKim) [reply .github/workflows/kind.yml:1]: yes automatically

File: test/e2e/case13_templatization_test.go (hunk 6 above)
Comment (yiraeChristineKim) [reply test/e2e/case13_templatization_test.go:109]: Also last spec

File: test/e2e/case6_no_ns_test.go (hunk 10 above)
Comment (yiraeChristineKim) [reply test/e2e/case6_no_ns_test.go:76]: These specs run as parallel so I added `clean up `to each test

File: test/e2e/case25_related_object_metric_test.go (hunk 11 above)
Comment (yiraeChristineKim) [reply test/e2e/case25_related_object_metric_test.go:69]: Thanks!!

File: test/resources/case20_delete_objects/case20_change_config_policy.yaml [hunk 13]
```diff
@@ -1,7 +1,7 @@
 apiVersion: policy.open-cluster-management.io/v1
 kind: ConfigurationPolicy
 metadata:
-  name: case20-name-changed
+  name: case20-name-changed-c20
```
[the remark this answers is not in the document]
Comment (mprahl) [reply test/resources/case20_delete_objects/case20_change_config_policy.yaml:4]: Perhaps just dropping `case20-` for it to match the others.

[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply test/resources/case20_delete_objects/case20_change_config_policy.yaml:4]: no good catch

File: test/e2e/case13_templatization_test.go [hunk 14]
```diff
@@ -194,9 +198,6 @@ var _ = Describe("Test templatization", func() {
 
 				return utils.GetComplianceState(managedPlc)
 			}, defaultTimeoutSeconds, 1).Should(Equal("Compliant"))
-			utils.Kubectl("delete", "configurationpolicy", case13CfgPolCreatePod, "-n", testNamespace)
```
[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply test/e2e/case13_templatization_test.go:197]: move to afterall

File: test/e2e/case20_delete_objects_test.go [hunk 15]
```diff
@@ -162,32 +161,31 @@ var _ = Describe("Test status fields being set for object deletion", func() {
 
 			deleteConfigPolicies(policies)
 
-			Consistently(func() interface{} {
+			utils.Kubectl("delete", "pod", case20PodName, "-n", "default", "--ignore-not-found")
+
+			Eventually(func() interface{} {
```
Comment (mprahl) [inline test/e2e/case20_delete_objects_test.go:166]: @JustinKuli what are your thoughts here about the change from `Consistently` to `Eventually`?

Comment (mprahl) [reply test/e2e/case20_delete_objects_test.go:166]: Actually, I didn't see that this was in the clean up so I think this should be fine.

File: test/resources/case20_delete_objects/case20_change_config_policy_not_prune.yaml [hunk 16]
```diff
@@ -0,0 +1,23 @@
+apiVersion: policy.open-cluster-management.io/v1
+kind: ConfigurationPolicy
+metadata:
+  name: case20-2-name-changed
+  namespace: managed
+spec:
+  remediationAction: enforce
+  pruneObjectBehavior: DeleteAll
```
Comment (mprahl) [inline test/resources/case20_delete_objects/case20_change_config_policy_not_prune.yaml:8]: Could `pruneObjectBehavior` be set to `None` or removed in this YAML and then remove the test code that patches the value now that this is a separate policy?

**Review by mprahl — COMMENTED** [review]
Just one minor comment but otherwise, this looks great to me! Nice job.

File: test/resources/case20_delete_objects/case20_change_config_policy_not_prune.yaml (hunk 16 above)
Comment (yiraeChristineKim) [reply test/resources/case20_delete_objects/case20_change_config_policy_not_prune.yaml:8]: good Thanks! I missed

Comment (yiraeChristineKim) [reply test/resources/case20_delete_objects/case20_change_config_policy_not_prune.yaml:8]: My purpose which created this yaml was separate the spec from "Clean up old object when configurationpolicy is changed" spec Thanks!!

**Comment by mprahl** [comment]
@yiraeChristineKim could you please squash the commits?

**Review by mprahl — APPROVED** [review]
